api/capi/src/nnstreamer-capi-tizen.c

```diff
@@ -0,0 +1,768 @@
+    gboolean marked = FALSE;
+
+    g_hash_table_iter_init (&iter, mm_handle->res_handles);
+    while (g_hash_table_iter_next (&iter, &key, &value)) {
+      pipeline_resource_s *mm_res = value;
+
+      if (mm_res->handle) {
+        mm_resource_manager_mark_for_release (mm_handle->rm_h, mm_res->handle);
+        mm_res->handle = NULL;
+        marked = TRUE;
+      }
+
+      if (destroy)
+        g_free (mm_res->type);
+    }
+
+    if (marked)
+      mm_resource_manager_commit (mm_handle->rm_h);
+  }
+
+  mm_resource_manager_set_status_cb (mm_handle->rm_h, NULL, NULL);
+  mm_resource_manager_destroy (mm_handle->rm_h);
+  mm_handle->rm_h = NULL;
+
+  mm_handle->invalid = FALSE;
+
+  if (destroy) {
+    g_hash_table_remove_all (mm_handle->res_handles);
+    g_free (mm_handle);
+  }
+}
+
+/**
+ * @brief Function to initialize mm resource manager.
+ */
+static int
+ml_tizen_mm_res_initialize (ml_pipeline_h pipe)
+{
+  ml_pipeline *p;
+  pipeline_resource_s *res;
+  tizen_mm_handle_s *mm_handle;
+  int status = ML_ERROR_STREAMS_PIPE;
+
+  p = (ml_pipeline *) pipe;
+
+  res = (pipeline_resource_s *) g_hash_table_lookup (p->resources, TIZEN_RES_MM);
+
+  /* register new resource handle of tizen mmfw */
+  if (!res) {
+    res = g_new0 (pipeline_resource_s, 1);
+    if (!res)
+      goto rm_error;
+
+    res->type = g_strdup (TIZEN_RES_MM);
+    g_hash_table_insert (p->resources, g_strdup (TIZEN_RES_MM), res);
+  }
+
+  mm_handle = (tizen_mm_handle_s *) res->handle;
+  if (!mm_handle) {
+    mm_handle = g_new0 (tizen_mm_handle_s, 1);
+    if (!mm_handle)
+      goto rm_error;
+
+    mm_handle->res_handles =
+        g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+
+    /* set mm handle */
+    res->handle = mm_handle;
+  }
+
+  status = ML_ERROR_NONE;
+
+rm_error:
+  return status;
+}
+
+/**
+ * @brief Function to acquire the resource from mm resource manager.
+ */
+static int
+ml_tizen_mm_res_acquire (ml_pipeline_h pipe,
+    mm_resource_manager_res_type_e res_type)
+{
+  ml_pipeline *p;
+  pipeline_resource_s *res;
+  tizen_mm_handle_s *mm_handle;
+  gchar *res_key;
+  int status = ML_ERROR_STREAMS_PIPE;
+  int err;
+
+  p = (ml_pipeline *) pipe;
+
+  res = (pipeline_resource_s *) g_hash_table_lookup (p->resources, TIZEN_RES_MM);
+  if (!res)
+    goto rm_error;
+
+  mm_handle = (tizen_mm_handle_s *) res->handle;
+  if (!mm_handle)
+    goto rm_error;
+
+  /* check invalid handle */
+  if (mm_handle->invalid)
+    ml_tizen_mm_res_release (mm_handle, FALSE);
+
+  /* create rm handle */
+  if (!mm_handle->rm_h) {
+    mm_resource_manager_h rm_h;
+
+    err = mm_resource_manager_create (MM_RESOURCE_MANAGER_APP_CLASS_MEDIA,
+        ml_tizen_mm_res_release_cb, pipe, &rm_h);
+    if (err != MM_RESOURCE_MANAGER_ERROR_NONE)
+      goto rm_error;
+
+    /* add state change callback */
+    err = mm_resource_manager_set_status_cb (rm_h, ml_tizen_mm_res_status_cb, pipe);
+    if (err != MM_RESOURCE_MANAGER_ERROR_NONE)
+      goto rm_error;
```

Comment:
I think `mm_resource_manager_destroy(rm_h)` is required because it's not assigned to `mm_handle->rm_h` yet.

Reply:
updated